### PR TITLE
Use a shader for low discrepancy label conversion

### PR DESCRIPTION
**Pull request by perlman**
# Description

The pull request will move the current label color calculation onto the GPU.

The current version of the code is from a few hours of pair session with @jni. It is a proof-of-concept showing that we (a) no longer need to convert the label values to [0,1] for vispy and (b) we _can_ do the computation within the shader.

Currently, only ISO categorical rendering is working, as shown in `examples/nD_labels.py`:
image

## Things to do before this is ready:
- [X] Implement shader for 3D label layers (ISO categorical)
- [X] Implement shader for 2D label layers
- [X] Fix the CPU-based color computation (for paint color preview & layer preview)
- [X] 3D iso categorical view
- [x] Test function of directly mapped colors 

## Not currently supported (for a future PR)
- [ ] Address uint64 labels. Labels are currently converted to float32
- [ ] Factor `seed` into a `uniform`

## Type of change
- [X] This change requires a documentation update

We should explicitly list the range of labels as those which can expressed as a float32.

# References

This is a small step towards addressing #204 (my initial suggestion for GPU-based color assignment).

The work in #2302 (2D fast labels) should be included in here and hopefully made redundant.

The current code for computation of low discrepancy images comes from #203 and #985.

# How has this been tested?
- [X] Additional tests added to confirm consistent coloring between GPU & CPU
- [X] Manual testing of the various examples (`examples/**labels*.py`)

## Final checklist:
- [ ] My PR is the minimum possible work for the desired functionality
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [X] I have added tests that prove my fix is effective or that my feature works
- [ ] If I included new strings, I have used `trans.` to make them localizable.
      For more information see our [translations guide](https://napari.org/docs/dev/guides/translations.html).

### CODE REVIEW

**Comment by kevinyamauchi** [comment]
I am stoked for this!

**Comment by alisterburt** [comment]
looks awesome @perlman ! 

Does this mean we would potentially be able to leverage linear interpolation without messing up labels or not really?

**Comment by sofroniewn** [comment]
@perlman any thoughts on revisiting this? There have been quite a few vispy releases since then - would be great to get this updated and over the finish line!! cc @jni

**Comment by perlman** [comment]
Yes; it's the agenda for working with @jni tomorrow.

**Comment by davidpross** [comment]
I am having some performance issues with multiscale labels in direct color mode with my application and on Zulip @jni directed me to this PR in progress. I tested my labels layer (with random colormap) in napari `main` branch and the `fast-labels-3` branch for this PR. Attached are the two pstat files from the profiling step. In each case I panned and zoomed manually in napari with the single layer loaded for one minute. So the steps will not be exactly the same but can hopefully be compared. I am not familiar with the logic for this rendering but one thing that jumps out is on the main branch the top function by total time is `colormap_utils.py:212(low_discrepancy_image)`.

[labels-pstat.zip](https://github.com/napari/napari/files/8789029/labels-pstat.zip)

**Comment by sofroniewn** [comment]
@perlman can you summarize what's still left to be done here/ what major roadblocks are. I see activity, but hard for me to get a sense for where we are and what's left. Thanks!! cc @potating-potato @jni

**Comment by jni** [comment]
Oh interesting, I was sure we'd written something @sofroniewn but in fact we've only just chatted repeatedly at the architecture meetings 😂.

The hold up is our support for arbitrary {label: color} mappings, which is hard to implement in a shader. The default random shader works which is pretty badass tbh!

@perlman maybe we should just add an ugly and slow fallback path that just falls back on the existing behaviour so that we can get this merged, and iterate in future PRs?

**Comment by perlman** [comment]
_[deleted -- verbose details for an efficient DIRECT color mode will go into a new issue or PR.]_

**Comment by perlman** [comment]
I had a misunderstanding about the behavior of DIRECT color mode for labels in napari.  For some reason, I thought the shader needed to do both pseudo-random colors *and* the dictionary mapped colors at the same time (e.g., for undefined colors). Looking at the current behavior, labels use pseudo random colors *or* dictionary defined values.

> @perlman maybe we should just add an ugly and slow fallback path that just falls back on the existing behavior so that we can get this merged, and iterate in future PRs?

@jni I think the ugly fallback path will be a bunch of if/else paths on `LabelColorMode.DIRECT` versus `LabelColorMode.AUTO` for things like scaling (or not scaling) values, setting random seed, etc.

The failing tests are highlighting where.

**Comment by sofroniewn** [comment]
> For some reason, I thought the shader needed to do both pseudo-random colors and the dictionary mapped colors at the same time (e.g., for undefined colors). Looking at the current behavior, labels use pseudo random colors or dictionary defined values.

Yeah just OR.

> @jni I think the ugly fallback path will be a bunch of if/else paths on LabelColorMode.DIRECT versus LabelColorMode.AUTO for things like scaling (or not scaling) values, setting random seed, etc.

I feel like shouldn't be too bad

> @perlman maybe we should just add an ugly and slow fallback path that just falls back on the existing behaviour so that we can get this merged, and iterate in future PRs?

But yes, lets try and get this in then asap and keep iterating in new PRs. I can review once tests all passing! Would be good to include some performance numbers at least in comments on PR before merge too. Will be great to get this in!!

**Comment by brisvag** [comment]
Pushed some changes that make 2D work! :)

Things still missing:
- [ ] properly handle unset labels (now they are ignored)
- [ ] dynamically set texture size depending on amoutn of labels?
- [x] hook up 3D
- [ ] get rid of old unused code

**Comment by jni** [comment]
@perlman and I just spent a bit more time playing with this. Here's what's broken currently, some things we'll need to fix in the PR, some things will be future issues:

- missing entries in the direct colormap are dealt with incorrectly. Sometimes there is some hash collision (or incorrectly found hash collision, probably through the linear chaining step) something and missing entries just take an adjacent color; other times, if it doesn't find a color, it horribly crashes the shader and I get this rather amusing error message:

```
2023-02-24 12:39:40.722 python[67867:45530852] GLDRendererMetal command buffer completion error: Error Domain=MTLCommandBufferErrorDomain Code=1 "Discarded (victim of GPU error/recovery) (00000005:kIOGPUCommandBufferCallbackErrorInnocentVictim)" UserInfo={NSLocalizedDescription=Discarded (victim of GPU error/recovery) (00000005:kIOGPUCommandBufferCallbackErrorInnocentVictim)}
```

Instead, missing keys should give a transparent color and be ignored in the iso categorical steps

- the translucent 3D labels rendering is broken completely
- the thumbnails are still broken, so the CPU code needs to be brought up to date
- probably for a totally separate issue/PR: when 3D textures are downsampled, the scale on the visual needs to be adjusted so that the coordinates are still correct. Currently the coordinates are off by the downsampling factor.

**Comment by jni** [comment]
> other times, if it doesn't find a color, it horribly crashes the shader

Ok yeah here's the issue: the GPU hash resolution doesn't check for the empty value:

https://github.com/perlman/napari/blob/3147c963525d2950b64a31d40f942888cb4ae21e/napari/_vispy/layers/labels.py#L66-L68

Compare with:

https://github.com/perlman/napari/blob/3147c963525d2950b64a31d40f942888cb4ae21e/napari/_vispy/layers/labels.py#L139-L142

So if the key is not in the table it's perfectly happy to march right off it into uncharted GPU memory 😂

**Comment by jni** [comment]
This is really close folks, and I'm excited!

First off, here's the difference in experience when painting between this PR and main for a large-ish labels image (~2500x2000) (with apologies to @jluethi's beautiful segmentation data 😂) (best viewed in full screen):

https://user-images.githubusercontent.com/492549/228791177-fb47b144-5cc0-409b-9f6f-b7e0e1911370.mov

It's not perfect, for that we'll need to drop the astype and just copy tiles directly, but it's just so much better! 😍 

Second, this fixes a crazy amount of bugs and performance issues we've seen over the years:
- I reckon we can close #204 (count those digits!!!!) with this — lots to optimise still but the brunt of it is now happening in the shader, not in the CPU!
- Closes #4384 (which you'll see has been hit by *many* people — it's a big one)
- Closes #4864 
- Closes #4942

I think the code still needs a bit of simplification, and a fair bit of testing, but **it works** 🎉 and I think I'd like some people to start playing around with it! CC @aeisenbarth @jluethi @Cryaaa 

btw current maximum number of colors in the color dict is 1e6 and performance would degrade quite a bit before reaching that... So, please keep that in mind. In principle, we can even bump that up quite a bit up to the maximum texture size.

The other limitation that this introduces is that labels are only accurate in [1, 2^23]. For values above that, the code works, but e.g. labels 2^24 and 2^24+1 would be aliased together. (ie we are limited by float32 precision for integers.) But I think that this is a worthwhile tradeoff, as it would only affect a small minority of users.

**Comment by perlman** [comment]
@jni & @brisvag -- this is a huge improvement! Thanks

A few questions:

> btw current maximum number of colors in the color dict is 1e6 and performance would degrade quite a bit before reaching that... So, please keep that in mind. In principle, we can even bump that up quite a bit up to the maximum texture size.

Where is the performance hit from? I thought we'd be fine as long as all values fit within a texture.

> The other limitation that this introduces is that labels are only accurate in [1, 2^23]. For values above that, the code works, but e.g. labels 2^24 and 2^24+1 would be aliased together. (ie we are limited by float32 precision for integers.) But I think that this is a worthwhile tradeoff, as it would only affect a small minority of users.

Clear next steps. =)

Finally, with `direct_lookup_shader`, how hard would it be to make `$use_selection` and `$selection` uniforms? When I last looked into it, I did not see an obvious clean way to do this from outside vispy, but I likely missed something.

**Comment by jni** [comment]
> Where is the performance hit from? I thought we'd be fine as long as all values fit within a texture.

Well, it's a hash table, so the higher the load, the higher the risk of collisions and the slower the performance. However, in the case of sequential labels under the hash table size, which indeed covers (I reckon) 99% of use cases, there would indeed be no performance penalty.

> Finally, with direct_lookup_shader, how hard would it be to make $use_selection and $selection uniforms? When I last looked into it, I did not see an obvious clean way to do this from outside vispy, but I likely missed something.

Oh yeah, I forgot to mention that — I moved *all* the "show selected" functionality to the CPU because it wasn't in the VisPy direct shader *and* it was interacting in weird ways with the auto shader. I don't think it would be hard to add back in, but it needed a fair bit of event-hooking-up so I decided to leave it to a later commit or later PR, not sure which yet. 😅 

And indeed I wanted to thank you @perlman and @brisvag for getting us this far — it's been a long road 😂 but we are so close and it's just gonna be a massive improvement for so many situations.

**Comment by brisvag** [comment]
> For values above that, the code works, but e.g. labels 2^24 and 2^24+1 would be aliased together.

You mean that the color looks the same? According to The Internet the human eye seems to distinguish between 1 and 10 million colors, and this is 16 mil... I think we're good xD

**Comment by jni** [comment]
LOL @brisvag it's worse than that: if you have a very large dataset with more than 2^24 labels, the issue is that some labels will *always have the same color*, regardless of what you put in the `layer.color` dictionary, because the label value ends up being the same in float32. So, this wouldn't work, even with two colors:

```python
labels = np.array([[2**24, 2**24+1]], dtype=np.int32)
viewer = napari.Viewer()
labels_layer = viewer.add_labels(labels)
labels_layer.color = {2**24: 'magenta', 2**24+1: 'green'}
```

Here you'd end up with both labels being magenta.

The only way around this is to relabel the data, which can be expensive for large datasets.

If I remember correctly we've already had people complain about needing such large labels (Yup: #2587 and #2620 — which I think can be closed actually, or at any rate would be closed here also!), so this is a real problem.

**Comment by GenevieveBuckley** [comment]
@jni pointed me in the direction of this PR. Since he [says this is really close now](https://github.com/napari/napari/pull/3308#issuecomment-1489997499), I'm trying ti figure out what is left remaining before it can be merged.

Can @perlman and @jni comment on this list summarizing remaining work? Thanks!

- [ ] *"Test function of directly mapped colors"* says Eric in the PR description. 
- [ ] Lots of tests were removed in `napari/layers/labels/_tests/test_labels.py`, presumably they will need to be added back in?
     - [ ] `test_large_custom_color_dict` 
     - [ ] `test_warning_too_many_colors`
     - [ ] `test_add_colors`
     - [ ] `test_switching_display_func`
     - [ ] `test_switching_display_func_during_slicing`
     - [ ] `test_add_large_colors`
- [ ] There are several `TODO` comments scattered about the code, presumably they are intended to be addressed in this PR? Were any intended to be left as future work instead?
    - [ ] `# TODO: should probably account for non-given labels` line 210 in vispy file `napari/_vispy/layers/labels.py` 
    - [ ] `# TODO: Trigger _selected_label mode in shader` line 818 in labels code `napari/layers/labels/labels.py`
    - [ ] `# TODO: Check not self.show_selected_label?` line 819 in labels code `napari/layers/labels/labels.py`
    - [ ] `# ASYNC_TODO: Do not compute the thumbnail until we are loaded. Is there a nicer way to prevent this from getting called?` line in labels code `napari/layers/labels/labels.py`
- [ ] Is there anything from [Juan's earlier summary of TODO items](https://github.com/napari/napari/pull/3308#issuecomment-1442680222) that hasn't been addressed?
- [ ] CI checks
    - [ ] Merge conflicts with main branch
    - [ ] ruff is failing in the pre-commit check
    - [ ] The CI is currently failing even before the test suite starts (but possibly fixing the merge conflicts and re-triggering CI will fix this?)

**Comment by jni** [comment]
> # TODO: Trigger _selected_label mode in shader line 818 in labels code napari/layers/labels/labels.py
> # TODO: Check not self.show_selected_label? line 819 in labels code napari/layers/labels/labels.py

Currently this happens in CPU. Probably simplest is to keep it that way and then move to shader in future PR.

**Comment by jni** [comment]
Agree with everything else on your list @GenevieveBuckley though!

**Comment by jni** [comment]
@GenevieveBuckley I should definitely be the one to do it since I reviewed all of the PRs that are causing issues now. Not unexpected. 😫

**Comment by perlman** [comment]
> @GenevieveBuckley I should definitely be the one to do it since I reviewed all of the PRs that are causing issues now. Not unexpected. 😫

I've fallen behind on keeping track of recent developments with the napari base layers, I'd gladly hand this to you @jni. :-P

@GenevieveBuckley: The list looks good to work from...

**Comment by GenevieveBuckley** [comment]
Thanks @jni - it's the conflict in the `_raw_to_displayed` function in `napari/layers/labels.py` that I'm having trouble figuring out.

**Comment by GenevieveBuckley** [comment]
The list of current TODO items for this PR is [in this comment](https://github.com/napari/napari/pull/3308#issuecomment-1598000991).

This is a list of future work items, for followup after this PR. Is there any thing extra that should be added here?

Future work:
- [ ] Address uint64 labels. Labels are currently converted to float32
- [ ] Factor `seed` into a `uniform`
- [ ] TODO: Trigger _selected_label mode in shader line 818 in labels code napari/layers/labels/labels.py
- [ ] TODO: Check not self.show_selected_label? line 819 in labels code napari/layers/labels/labels.py
- [ ] *?The other limitation that this introduces is that labels are only accurate in [1, 2^23]. For values above that, the code works, but e.g. labels 2^24 and 2^24+1 would be aliased together. (ie we are limited by float32 precision for integers.) But I think that this is a worthwhile tradeoff, as it would only affect a small minority of users."*
    >>> Clear next steps. =)
    >>
    >> You mean that the color looks the same? According to The Internet the human eye seems to distinguish between 1 and 10 million colors, and this is 16 mil... I think we're good xD https://github.com/napari/napari/pull/3308#issuecomment-1494011833
    >
    > LOL @brisvag it's worse than that: if you have a very large dataset with more than 2^24 labels, the issue is that some labels will always have the same color, regardless of what you put in the layer.color dictionary, because the label value ends up being the same in float32. So, this wouldn't work, even with two colors:

**Comment by jluethi** [comment]
This PR is awesome!

In my old setups (e.g. in a napari 4.16 env used for https://github.com/napari/napari/issues/4384 testing), I got these patterns when using direct color mode and trying to show a continuous viridis gradient left to right:
napari_416_shuffled

With this branch, all the colors are displayed correctly:
fast_labels_shuffled

That's 500k labels being colored correctly here. For reference, this is the label image with auto color mode:
Screenshot 2023-07-21 at 15 56 07

So big ❤️ @jni @perlman @GenevieveBuckley @brisvag & everyone else! Great upgrade to make label display robust 👏🏻


<details>
<summary>Code to recreate this</summary>
import matplotlib
import numpy as np
import pandas as pd
import napari

# Set the number of steps
nb_steps = 5000
y_size = 10
rows = 100

# Create a dummy label image
base = np.linspace(start=1, stop=nb_steps, num=nb_steps).astype('uint32')
# label_img_base = np.repeat(base.reshape([1, base.shape[0]]), y_size, axis=0)
linspace_list = []
# df = pd.DataFrame()
colormap_all = {}
for row in range(rows):
    print(row)
    shifted_img = base + row * nb_steps
    np.random.shuffle(shifted_img)
    label_img = np.repeat(shifted_img.reshape([1, base.shape[0]]), y_size, axis=0)
    linspace_list.append(label_img)
    df_new = pd.DataFrame([label_img[0, :], base]).T
    df_new.columns = ['label', 'feature']
    colormap_features = (df_new['feature'] - min(df_new['feature'])) / (max(df_new['feature']) - min(df_new['feature']))
    colors_ordered = matplotlib.colormaps['viridis'](colormap_features)
    colormap_ordered = dict(zip(df_new['label'].astype(int), colors_ordered))
    colormap_all.update(colormap_ordered)
label_img = np.concatenate(linspace_list)

# Add to napari
viewer = napari.Viewer()
viewer.add_image(label_img, colormap='viridis')
labels_layer_ordered = viewer.add_labels(label_img, opacity=100)

# Set the label image colormaps
labels_layer_ordered.color = colormap_all
viewer.show(block=True)
</details>

**Review by brisvag — APPROVED** [review]
